pkg/apis/networking/validation/validation.go

```diff
@@ -385,33 +396,64 @@ func validateHTTPIngressPath(path *networking.HTTPIngressPath, fldPath *field.Pa
 		}
 	}
 
-	allErrs = append(allErrs, validateIngressBackend(&path.Backend, fldPath.Child("backend"), opts)...)
+	allErrs = append(allErrs, validateIngressBackend(&path.Backend, fldPath.Child("backend"), opts, requestGV)...)
 	return allErrs
 }
 
+// portFieledName returns the name of the field used for an
+// ingress service port in the provided GroupVersion.
+func portFieldName(gv schema.GroupVersion) string {
+	switch gv {
+	case networkingv1beta1.SchemeGroupVersion, extensionsv1beta1.SchemeGroupVersion:
+		return "servicePort"
+	default:
+		return "port"
+	}
+}
+
+// serviceName returns the name of the field used for a
+// service in the provided GroupVersion.
+func serviceName(gv schema.GroupVersion) string {
+	switch gv {
+	case networkingv1beta1.SchemeGroupVersion, extensionsv1beta1.SchemeGroupVersion:
+		return "serviceName"
+	default:
+		return "name"
+	}
+}
+
 // validateIngressBackend tests if a given backend is valid.
-func validateIngressBackend(backend *networking.IngressBackend, fldPath *field.Path, opts IngressValidationOptions) field.ErrorList {
+func validateIngressBackend(backend *networking.IngressBackend, fldPath *field.Path, opts IngressValidationOptions, requestGV schema.GroupVersion) field.ErrorList {
 	allErrs := field.ErrorList{}
 
 	hasResourceBackend := backend.Resource != nil
-	hasServiceBackend := len(backend.ServiceName) > 0 || backend.ServicePort.IntVal != 0 || len(backend.ServicePort.StrVal) > 0
+	hasServiceBackend := backend.Service != nil
 
 	switch {
 	case hasResourceBackend && hasServiceBackend:
 		return append(allErrs, field.Invalid(fldPath, "", "cannot set both resource and service backends"))
 	case hasResourceBackend && !opts.allowResourceBackend:
 		return append(allErrs, field.Forbidden(fldPath.Child("resource"), "not supported; only service backends are supported in this version"))
-	case hasResourceBackend:
-		allErrs = append(allErrs, validateIngressTypedLocalObjectReference(backend.Resource, fldPath.Child("resource"))...)
-	default:
-		if len(backend.ServiceName) == 0 {
-			return append(allErrs, field.Required(fldPath.Child("serviceName"), ""))
+	case hasServiceBackend:
```

Comment:
structure this to validate service name and port individually, and to accumulate as many errors as possible:

```go
case hasServiceBackend:

  if len(backend.Service.Name) == 0 {
    allErrs = append(allErrs, field.Required(serviceNameFieldPath(fldPath, requestGV), ""))
  } else {
    for _, msg := range apivalidation.ValidateServiceName(backend.ServiceName, false) {
      allErrs = append(allErrs, field.Invalid(serviceNameFieldPath(fldPath, requestGV), backend.Service.Name, msg))
    }
  }

  hasPortName := len(backend.Service.Port.Name) > 0
  hasPortNumber := backend.Service.Port.Number != 0
  if hasPortName && hasPortNumber {
    allErrs = append(allErrs, ... error that only one of port name and number can be specified
  } else if hasPortName {
    for _, msg := range validation.IsValidPortName(backend.Service.Port.Name) {
      allErrs = append(allErrs, field.Invalid(servicePortNameFieldPath(fldPath, requestGV), backend.Service.Port.Name, msg))
    }
  } else if hasPortNumber {
    for _, msg := range validation.IsValidPortNum(int(backend.Service.Port.Number)) {
      allErrs = append(allErrs, field.Invalid(servicePortNumberFieldPath(fldPath, requestGV), backend.Service.Port.Number, msg))
    }
  } else {
    allErrs = append(allErrs, ... error that one of port name and number is required
  }
```